src/cache.rs

```diff
@@ -0,0 +1,305 @@
+
+/// If the env var `ARECIBO_CACHE` exists, then use that value.
+/// Otherwise, set `ARECIBO_CACHE` to be `$HOME/.arecibo/`.
+fn init_cache() -> Cache {
+  let path = if let Ok(x) = std::env::var("ARECIBO_CACHE") {
+    tracing::debug!("{:?}", &x);
+    PathBuf::from(x)
+  } else {
+    let path = home::home_dir()
+      .expect("targets without $HOME not supported")
+      .join(".arecibo/");
+    std::env::set_var("ARECIBO_CACHE", &path);
+    path
+  };
+
+  create_dir_all(&path).unwrap();
+  create_dir_all(path.join("commitment_keys")).unwrap();
+  create_dir_all(path.join("public_params")).unwrap();
+
+  Cache {
+    inner: Arc::new(Mutex::new(path)),
+  }
+}
+
+/// Cache structure that holds the root directory of all the cached files
+pub struct Cache {
+  inner: Arc<Mutex<PathBuf>>,
+}
+
+impl Cache {
+  /// Sets the inner root directory of the cache
+  pub fn set_inner(&self, path: &PathBuf) {
+    self.inner.lock().unwrap().clone_from(path);
+
+    create_dir_all(path).unwrap();
+    create_dir_all(path.join("commitment_keys")).unwrap();
+    create_dir_all(path.join("public_params")).unwrap();
+  }
+
+  /// Gets the inner root directory of the cache
+  pub fn get_inner(&self) -> PathBuf {
+    self.inner.lock().unwrap().to_path_buf()
+  }
+
+  /// Returns the commitment key file, and creates one if it doesn't exist
+  pub fn ck_file<G: Group>(&self) -> io::Result<File> {
+    let path = self.get_inner().join("commitment_keys").join(G::name());
+    let exists = path.exists();
+
+    let mut file = OpenOptions::new()
+      .read(true)
+      .write(true)
+      .create(true)
+      .open(path)?;
+
+    if !exists {
+      let ck = G::CE::setup(b"ck", 0);
+      file.write_u64::<LittleEndian>(ck.length() as u64)?;
+      ck.encode(&mut file)?;
+    }
+
+    file.rewind()?;
+    Ok(file)
+  }
+
+  /// Returns the file of the given digest, and creates one if it doesn't exist
+  pub fn digest_file<F: PrimeField>(&self, digest: F) -> io::Result<(File, bool)> {
+    let path = self
+      .get_inner()
+      .join("public_params")
+      .join(format!("{:?}", digest));
+    let exists = path.exists();
+
+    let mut file = OpenOptions::new()
+      .read(true)
+      .write(true)
+      .create(true)
+      .open(path)?;
+
+    file.rewind()?;
+    Ok((file, exists))
+  }
+
+  /// Returns a commitment key from the cache. If the cached commitment key is shorter
+  /// than what we need, we extend the key and update the cache as a side effect.
+  ///
+  /// This is the same as calling `G::CE::setup(b"ck", n)`
+  pub fn commitment_key<G: Group>(&self, n: usize) -> io::Result<CommitmentKey<G>> {
+    let mut file = self.ck_file::<G>()?;
+
+    let ck_size = file.read_u64::<LittleEndian>()? as usize;
+    println!("ck_size: {}, n: {}", ck_size, n);
+
+    let mut mmap = unsafe { MmapMut::map_mut(&file)? };
+    let skip = std::mem::size_of::<usize>();
+    let bytes = &mut mmap[skip..]; // skip the first `usize` bytes
```

Comment:
Ah, so we should at least make clear that despite the use of concurrency-safe structures above, this is *not* concurrency-safe, right? Two calls to `commitment_key` (including those generated as simply as running unit tests concurrently) could easily step on each other's toes.

Have you considered something like a `struct MemoryMappedMutFile(Arc<RwLock<MmapMut>>);` of which the cache would create a singleton instance per relevant file (commitment key, public params)?